tlair/src/main/scala/at/forsyte/apalache/tla/lir/TlaDeclLevelFinder.scala

```diff
@@ -0,0 +1,55 @@
+package at.forsyte.apalache.tla.lir
+
+/**
+ * This class computes the level of a TLA+ declaration by using TlaExLevelFinder. See Lamport. Specifying Systems, p. 322.
+ *
+ * @param module a TLA module that contains the declarations to query
+ * @author Igor Konnov
+ */
+class TlaDeclLevelFinder(module: TlaModule) {
+  private val consts = Set(module.constDeclarations.map(_.name): _*)
+  private val vars = Set(module.varDeclarations.map(_.name): _*)
+  private val defs = Map(module.operDeclarations map { d => d.name -> d }: _*)
+  private var levelCacheMap: Map[String, TlaLevel] = Map()
+
+  def apply(decl: TlaDecl): TlaLevel = {
+    decl match {
+      case TlaConstDecl(_) =>
+        TlaLevelConst
+
+      case TlaVarDecl(_) =>
+        TlaLevelState
+
+      case TlaAssumeDecl(_) =>
+        TlaLevelConst
+
+      case TlaOperDecl(name, _, _) =>
+        cacheLevel(Set(name), name)
+        levelCacheMap(name)
```

Comment:
Given that we are caching results, why don't we first check whether the `name` is cached before we try to add it to the cache?

Reply:
yep, added that